File: Daily-Streak/Day-09_Smallest-Subtree-with-All-the-Deepest-Nodes/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    // Helper class to store depth and subtree node
    private static class Pair {
        int depth;
        TreeNode node;

        Pair(int depth, TreeNode node) {
            this.depth = depth;
            this.node = node;
        }
    }

    private Pair dfs(TreeNode root) {
        if (root == null) {
            return new Pair(0, null);
        }

        Pair left = dfs(root.left);
        Pair right = dfs(root.right);

        if (left.depth > right.depth) {
            return new Pair(left.depth + 1, left.node);
        } else if (right.depth > left.depth) {
            return new Pair(right.depth + 1, right.node);
        } else {
            return new Pair(left.depth + 1, root);
        }
    }

    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        // ✅ OPTIMAL DFS APPROACH (ACTIVE)
        return dfs(root).node;
    }

    /*
    ===========================
    ❌ BRUTE FORCE IDEA (COMMENTED)
    ===========================
    - Track depths using map
    - Find deepest nodes
    - Compute LCA separately

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(5);
        root.right = new TreeNode(1);
        root.left.left = new TreeNode(6);
        root.right.right = new TreeNode(8);

        Solution s = new Solution();
        System.out.println(s.subtreeWithAllDeepest(root).val);
    }
    */
}

File: Daily-Streak/Day-09_Smallest-Subtree-with-All-the-Deepest-Nodes/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right)
 *         : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Helper function returns pair:
    // {maximum depth from this node, subtree root containing deepest nodes}
    pair<int, TreeNode*> dfs(TreeNode* root) {
        if (!root) {
            return {0, nullptr};
        }

        auto left = dfs(root->left);
        auto right = dfs(root->right);

        // If left subtree is deeper
        if (left.first > right.first) {
            return {left.first + 1, left.second};
        }
        // If right subtree is deeper
        else if (right.first > left.first) {
            return {right.first + 1, right.second};
        }
        // If both sides have equal depth
        else {
            return {left.first + 1, root};
        }
    }

    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        // ✅ OPTIMAL DFS APPROACH (ACTIVE)
        return dfs(root).second;
    }

    /*
    ===========================
    ❌ BRUTE FORCE IDEA (COMMENTED)
    ===========================
    - Find all deepest nodes
    - Compute LCA of all deepest nodes
    - Multiple traversals + extra storage

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    #include <bits/stdc++.h>
    using namespace std;

    int main() {
        TreeNode* root = new TreeNode(3);
        root->left = new TreeNode(5);
        root->right = new TreeNode(1);
        root->left->left = new TreeNode(6);
        root->right->right = new TreeNode(8);

        Solution s;
        TreeNode* ans = s.subtreeWithAllDeepest(root);
        cout << ans->val << endl;
        return 0;
    }
    */
};
